python/ray/rllib/rollout.py

```diff
@@ -106,14 +106,13 @@ def rollout(agent, env_name, num_steps, out=None, no_render=True):
         env = gym.make(env_name)
 
     if hasattr(agent, "local_evaluator"):
-        state_init = agent.local_evaluator.policy_map[
-            "default"].get_initial_state()
```

Comment:
I'm kind of wondering if it makes sense to instead call local_evaluator.sample() in complete episodes mode. That will generate complete episodes trajectories if batch_mode is set to complete_episodes in the agent config.

Any downsides you see of using that for rollout instead of the existing code below?

Reply:
Hum, not familiar with that part yet. I’ll have a look.

Reply:
@ericl the downside I see is when agent has no `local_evaluator` attribute, then you still need the existing code to generate trajectories.

Reply:
well, shouldn't be a problem

Reply:
I'm facing a weird problem: I'm testing against a 3 agents env with agents having same episode length. Every X time steps, `env.step()` returns `{'agent1': True, 'agent2': True, 'agent3': True}` for done

However when I call `agent.local_evaluator.sample()` and iterate through `policy_batches`, I see that each policy have 3 times True value in dones. See an example below:

```
agent_1 SampleBatch({'obs': array([[ 0.  ,  0.3 ,  1.  , ..., 18.5 ,  1.  , 14.  ],
       [ 0.  ,  0.6 ,  1.1 , ..., 18.75,  1.  , 13.  ],
       [ 0.  ,  0.7 ,  1.2 , ..., 19.  ,  1.  , 12.  ],
       ...,
       [ 1.  , -2.8 , -2.8 , ..., 14.25,  1.  , 31.  ],
       [ 1.  , -2.8 , -2.8 , ..., 14.5 ,  1.  , 30.  ],
       [ 1.  , -2.8 , -2.8 , ..., 14.75,  1.  , 29.  ]], dtype=float32), 'actions': array([ 6, 30, 30,  9, 29, 29, 29, 38, 30, 38,  1, 29, 68,  1,  2,  2, 10,
        1,  9,  2,  2,  2,  2,  2,  9,  9,  9,  9,  9,  9,  9,  6,  9,  9,
        9,  6,  6,  6, 28, 28,  6,  9,  6,  6, 28, 28, 28, 28, 16, 28]), 'rewards': array([-4.304386 , -4.6716704, -5.14568  , -5.33315  , -5.467442 ,
       -5.4712877, -5.5904975, -5.7080307, -5.6359763, -5.6135325,
       -5.770061 , -5.594026 , -4.055325 , -1.9725   , -6.2327714,
       -6.2494674, -5.9769073, -5.846483 , -5.7549148, -5.584145 ,
       -5.41025  , -5.1662507, -4.7459483, -4.230974 , -3.8007362,
       -3.6591623, -3.6169324, -3.6091404, -3.6216927, -3.5994544,
       -3.5712452, -3.5403967, -3.5260623, -3.5144002, -3.4872878,
       -3.4654713, -3.469382 , -3.5050275, -3.5869343, -3.579932 ,
       -3.548359 , -3.4565392, -3.5001438, -3.5023057, -3.527714 ,
       -3.5065172, -3.6021357, -3.6020231, -2.61638  , -1.3808   ],
      dtype=float32), 'new_obs': array([[ 0.  ,  0.8 ,  1.4 , ..., 19.25,  1.  , 11.  ],
       [ 0.  ,  1.  ,  1.5 , ..., 19.5 ,  1.  , 10.  ],
       [ 0.  ,  1.1 ,  1.7 , ..., 19.75,  1.  ,  9.  ],
       ...,
       [ 1.  , -2.8 , -2.8 , ..., 15.  ,  1.  , 28.  ],
       [ 1.  , -2.8 , -2.8 , ..., 15.  ,  1.  , 28.  ],
       [ 1.  , -2.8 , -2.8 , ..., 15.  ,  1.  , 28.  ]], dtype=float32), 'dones': array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False])
```

Shouldn't there be only 1 True per policy batch?

Reply:
The reason was using DQN with n_step = 3 (hence trajectory post processing applies the same done value to 3 steps). So no way to have clear view on terminated episodes (unless considering consecutive dones a single one? but that's probably too fragile). 
This is only a problem when using batch_mode = truncate_episodes since a batch can contain multiple episodes. With complete_episodes no need to care about where dones are to compute episodes reward.
What would be the best option for you @ericl? Treat complete and truncate differently in rollout.py? Or maybe force use of complete_episode when using it?

Reply:
Well, given that some algorithms don't have policy_evaluator (i.e., ES, ARS), and that postprocessing can transform the trajectories as you've found, I think it may be better to keep the code for rollout and evaluation separate.

That way we don't need to worry about episode postprocessing. What do you think?